crates/nu-engine/src/shell/shell_manager.rs

```diff
@@ -16,13 +18,32 @@ use std::sync::Arc;
 pub struct ShellManager {
     pub current_shell: Arc<AtomicUsize>,
     pub shells: Arc<Mutex<Vec<Box<dyn Shell + Send>>>>,
+
+    /// Windows-specific: keep track of previous cwd on each drive across fs_shell instances
+    #[cfg(windows)]
+    windows_drives_previous_cwd: Arc<Mutex<IndexMap<String, String>>>,
 }
 
 impl ShellManager {
     pub fn enter_script_mode(&self) -> Result<(), std::io::Error> {
         //New fs_shell starting from current path
-        let fs_shell = FilesystemShell::with_location(self.path(), FilesystemShellMode::Script)?;
-        self.insert_at_current(Box::new(fs_shell));
+        let mut fs_shell =
+            FilesystemShell::with_location(self.path(), FilesystemShellMode::Script)?;
+        self.add_windows_drives_previous_cwd_to_shell(&mut fs_shell);
```

Comment:
We share the one instance of add_windows_drives_previous_cwd with all fs_shells. This allows to use this feature accross entered fs shells.